host/rpc.go

```diff
@@ -0,0 +1,1017 @@
+	if initial.WindowEnd > existing.Revision.WindowEnd {
+		extension := initial.WindowEnd - existing.Revision.WindowEnd
+		baseStorageCost = settings.StoragePrice.Mul64(initial.Filesize).Mul64(extension)
+		baseCollateral = settings.Collateral.Mul64(initial.Filesize).Mul64(extension)
+	}
+
+	additionalCollateral := initial.TotalCollateral.Sub(baseCollateral)
+	validHostValue := settings.ContractFee.Add(baseStorageCost).Add(initial.TotalCollateral)
+	missedHostValue := settings.ContractFee.Add(baseStorageCost).Add(additionalCollateral)
+
+	// calculate the amount the host needs to fund as the difference between the
+	// valid proof output, base storage revenue, and contract fee. The renter is
+	// responsible for funding the remainder.
+	if initial.HostOutput.Value.Cmp(validHostValue) != 0 {
+		err := fmt.Errorf("renewal %v rejected: expected %v host value, received %v", existingID, validHostValue, initial.HostOutput.Value)
+		log.Warnln(err)
+		r.writeErr(err)
+		return err
+	} else if initial.MissedHostValue.Cmp(missedHostValue) != 0 {
+		err := fmt.Errorf("renewal %v rejected: expected %v host missed value, received %v", existingID, missedHostValue, initial.MissedHostValue)
+		log.Warnln(err)
+		r.writeErr(err)
+		return err
+	} else if initial.TotalCollateral.Cmp(settings.MaxCollateral) > 0 {
+		err := fmt.Errorf("renewal %v rejected: host max collateral is %v, expected to post %v", existingID, settings.MaxCollateral, initial.TotalCollateral)
+		log.Warnln(err)
+		r.writeErr(err)
+		return err
+	}
+
+	// sign the renewal and finalization.
+	renewalTxn.FileContractResolutions[0].Renewal.InitialRevision.HostSignature = r.privkey.SignHash(initialSigHash)
+	renewalTxn.FileContractResolutions[0].Renewal.FinalRevision.HostSignature = r.privkey.SignHash(finalRevisionSigHash)
+	renewalSigHash := vc.RenewalSigHash(renewalTxn.FileContractResolutions[0].Renewal)
+	renewalTxn.FileContractResolutions[0].Renewal.HostSignature = r.privkey.SignHash(renewalSigHash)
+
+	renterInputs, renterOutputs := len(renewalTxn.SiacoinInputs), len(renewalTxn.SiacoinOutputs)
+	toSign, cleanup, err := r.wallet.FundTransaction(&renewalTxn, initial.TotalCollateral, r.tpool.Transactions())
+	if err != nil {
+		err = fmt.Errorf("failed to fund renewal transaction: %w", err)
+		log.Warnln(err)
+		// should not send the actual error to the renter, just a generic message.
+		r.writeErr(errors.New("failed to renew contract"))
+		return err
+	}
+	defer cleanup()
+
+	// send the renter the host's unsigned inputs, additional outputs, and
+	// contract signatures.
+	hostAdditions := &rhp.RPCRenewContractHostAdditions{
+		Inputs:  renewalTxn.SiacoinInputs[renterInputs:],
+		Outputs: renewalTxn.SiacoinOutputs[renterOutputs:],
+		// TODO: should hosts use rollover? Currently assumes that the host does
+		// not want to rollover anything into the contract and should use new funds.
+		HostRollover:          types.ZeroCurrency,
```

Comment:
I think there's a little more nuance for hosts since it reduces their visible payouts. The host may see no revenue until the renter stops using the host. It's mathematically equivalent but might be hard to understand from a user perspective.

Reply:
true. Maybe we could cap it at the collateral for the *current* contract? That way, we can display it to hosts as:
```
Contract profit:        X SC
Collateral returned:    Y SC
Collateral rolled over: Z SC
```
I think hosts will be less likely to complain about this, since collateral is already perceived as being "locked up"/inaccessible.